### ISSUE TITLE
mongo-connector stuck on fatal error

### ISSUE THREAD

**Issue by joe-mojo** [mongodb-labs/mongo-connector#279]
I have 2 replicaset in 2 different environnment (dev&prod) with no issue. They run mongo-connector 2.0.3 with pyMongo 2.8. Prod is running with python 2.6, dev with 2.7.

I tried recently to set up a 3thd env with dev data. I ran into fatal issue. This occured with pyMongo 2.0.3, 2.10 and pyMongo 2.8 and 2.8.1.

When starting mongo-connector service, after a constant number of 6000 doc inserted into ElasticSearch. The 6001nth is raising: 
```
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/local/lib/python2.7/dist-packages/mongo_connector/util.py", line 85, in wrapped
    func(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/mongo_connector/oplog_manager.py", line 185, in run
    for n, entry in enumerate(cursor):
  File "/usr/local/lib/python2.7/dist-packages/pymongo/cursor.py", line 1090, in next
    if len(self.__data) or self._refresh():
  File "/usr/local/lib/python2.7/dist-packages/pymongo/cursor.py", line 1039, in _refresh
    limit, self.__id))
  File "/usr/local/lib/python2.7/dist-packages/pymongo/cursor.py", line 958, in __send_message
    self.__compile_re)
  File "/usr/local/lib/python2.7/dist-packages/pymongo/helpers.py", line 121, in _unpack_response
    compile_re)
  File "/usr/local/lib/python2.7/dist-packages/bson/__init__.py", line 537, in decode_all
    tz_aware, uuid_subtype, compile_re))
  File "/usr/local/lib/python2.7/dist-packages/bson/__init__.py", line 331, in _elements_to_dict
    data, position, as_class, tz_aware, uuid_subtype, compile_re)
  File "/usr/local/lib/python2.7/dist-packages/bson/__init__.py", line 320, in _element_to_dict
    data, position, as_class, tz_aware, uuid_subtype, compile_re)
  File "/usr/local/lib/python2.7/dist-packages/bson/__init__.py", line 159, in _get_object
    encoded, as_class, tz_aware, uuid_subtype, compile_re)
  File "/usr/local/lib/python2.7/dist-packages/bson/__init__.py", line 331, in _elements_to_dict
    data, position, as_class, tz_aware, uuid_subtype, compile_re)
  File "/usr/local/lib/python2.7/dist-packages/bson/__init__.py", line 320, in _element_to_dict
    data, position, as_class, tz_aware, uuid_subtype, compile_re)
  File "/usr/local/lib/python2.7/dist-packages/bson/__init__.py", line 232, in _get_date
    dt = EPOCH_AWARE + datetime.timedelta(seconds=seconds)
InvalidBSON: date value out of range
```
If I set timeZoneAware to false, this is exactly the same, expect `EPOCH_AWARE`is replaced by `EPOCH_NAIVE`.

So you think the 6001nth record I imported is bad ? Nope. The first time I launched mongo-connector, I have already imported 180,000+ records, but the second time I juste droped the collection and re-imported all of them so that mongo-connector started to insert doc in ES at the time I start import. Same issue. Setting noDump=true had no effect. The proof the 6001nth doc is not bad ? By spliting import file into 2 parts, (1: 6000 records, 2: remaining), I got 6000 record then 100 imported before first error. 6000 and 100 where constants. By splitingn one more time (6000+6000+remaining), I was hable to import 6000+283 docs. Strange. So pyMongo was not detectin a bad BSON, there is a kind of race condition underneath.

continueOnError and batchSize have no effect, each time I restart mongo-connector without removing oplog-timestamp, even with a dropped collection, it will try to send to ES the same docs again and again. I expected batchSize=1 to force mongo-connector to have an up to date timestamp, but this is not working. the last few records before the fatal error are always "repeated".

The bad part is that I'm not able to launch mongo-connector only to refresh timestamp and exiting, is is crashing very fast and do not update anything. I'm still looking for a workaround : my goal start from an empty collection, start mongo-connector that will remain quiet, then starting to insert doc at a slow rate to see if it matters.

Please upgrade pyMongo dep if it solve this solution (pyMongo 3 is supposed to have a better date support), or forward the issue to pyMongo if the latest has the same issue.
Anyway, you should fix mongo-connector so that oplog-timestamp is updated to avoid doc manager insertion replay.

**Comment by llvtt** [comment]
This looks like PyMongo is having trouble decoding a date in an oplog record. This error is occurring while tailing the oplog, so it has nothing to do with the collection dump, which is what `noDump` and `continueOnError` concern. mongo-connector updates its timestamp file with the *latest* timestamp from the last batch of oplog entries it was able to process, so it's not possible to skip oplog entries without modifying the timestamp file manually.

What is the "import file" that you're referring to, and what do you mean by splitting it up? Are you using `mongoimport` to import data to MongoDB that you're then synchronizing to Elasticsearch with mongo-connector?

It would be useful to try and isolate the offending record to find out more about what's going wrong.

**Comment by joe-mojo** [comment]
Thx for this for answering so quickly !
Yes, on each tryout I'm using mongoimport. Before identifying it was an oplog issue, I ran mongoimport then mongo-connector, but following tryouts were done by lauching mongo-connector then mongoimport.

> What is the "import file" that you're referring to, and what do you mean by splitting it up?

My import file come from mongoexport. It is the usual one-json-object-per-line file.

> It would be useful to try and isolate the offending record to find out more about what's going wrong

As I explained above, there is no offending entry. As far as I can see by splitting the import file, the record on which mongo-connector is constant for a file split only. If I retry everything (drop collection, restart connector, re-import file by file) with a different split, the offending record is not the same !
See : 
 - 1 file with 180,000+ records : always stuck at 6000. Removing the 6001nth doesn't help.Even multiple times.
 - 2 files : (6000, 174,000+) : always stuck at 6100
 - 3 files : (6000, 6000, 168, 000+) : always stuck at 6283
So I didn't tried other splitting. 

How do you explain the "offending record" depends on how much file I import and how big they are ?

> mongo-connector updates its timestamp file with the latest timestamp from the last batch of oplog entries it was able to process

So there is another bug : even a batch size of 1 doesn't ensure that the oplog-timestamp is updated after processing every record.

**Comment by llvtt** [comment]
mongo-connector updates the oplog progress file once per second with with mongo-connector's internal timestamp. The internal timestamp is updated per `batchSize`. The purpose of the internal timestamp is for recreating the cursor at the correct position in the oplog, in case the cursor is destroyed for any reason.

I can see how this could be confusing. I'll file an issue about refactoring the way that updating the timestamp file works, so that it gets updated whenever the internal timestamp is updated. I don't see any benefit in always sticking to a "once per second" rule.

Nonetheless, it's intentional that mongo-connector doesn't allow you to skip oplog records. If you could, then future oplog entries might not make any sense. For example, you might skip the insert of a document that gets updated later, but that update doesn't make any sense without the original document being there.

I'm certain that there's at least one record in your file that causes PyMongo to raise InvalidBSON. mongo-connector itself never raises this Exception type. If you want to find out exactly what record this is, try inserting one record at a time until the Exception is thrown. When it is, you can get the oplog record that caused the Exception using the `mongo` shell (or the equivalent in any driver):

```javascript
> use local
> db.oplog.rs.count()
<some number, call it X>
> db.oplog.rs.find().skip(X - 1).limit(1).pretty()
<offending oplog record is printed>
```
If you really, really want to skip the offending oplog entry, you need to find the Timestamp of the record that comes after it. This requires knowing what oplog entry is causing the problem in the first place. Once you find the Timestamp of the record that comes after the offending one, you need to edit the oplog progress file (called `oplog.timestamp` by default). Check out the [wiki page](https://github.com/10gen-labs/mongo-connector/wiki/Oplog-Progress-File#format) for the format of the file.

**Comment by joe-mojo** [comment]
Ok, I'll find the offending oplog record. But the issue is not that
mongo-connector is not updatin oplog-timestamp once per second. The issue
is that the timestamp is NOT updated every record even if I set batchSize=1.

Today, I tried the following workarounds :
 - restart my replicaset from an empty DB to have no oplog -> oplog is
polluted by some records because mongo-connector is crashiongwithe the same
error
 - Do the same, but after cleaning the import file. -> more records sent to
ES, but crash
Offending recors may be date in the far future (I discovered our API made
no control so you could record a birthDate in the future, say 19987 instead
of 1987). When you are more than 10000 years from now, my mongo client
(robomongo) doesn't display date value as an ISO string but as a timestamp,
and mongoexport 3.0 export it as {$date: { $numberLong: 12345... }} instead
of {$date: "iso string"}.

I don't know if tose far-future date are offending is present in oplog, but
I know for sure that the fresh new Java mongodb driver don't understand
{$date: { $numberLong: 12345... }} thought it has been exported by
mongoexport 3.0.

So that's a new bug discovered by the way...

The final workaround was replacing an instance in the replicaset (debian
wheezy) by a prod clone (ubuntu 10.04) where mongo-connector is working
like a charm. The bug may be caused by some king of install, as dev
instance ((debian wheezy too) are running mongo-connector without problem.
The instance having the date bug where installed with some dev package as
at first install, mongo-connector was warning they where missin, so some
python modules wouldn't be compiled.

So, in a nutshell
 - installation with full compilation may cause the date bug in BSON
dependency of pyMongo (to be confirmed)
 - mongo-connector is not using the last version of pyMongo
 - batchSize doesn't behave as expected
 - up-to-date Java driver is not consistent with up-to-date mongoexport
output.

2015-07-10 20:10 GMT+02:00 Luke Lovett <notifications@github.com>:

**Comment by ilooch** [comment]
I have the same issue.
The mongo-connector stop working and I found the same exception as @joe-mojo 
>InvalidBSON: date value out of range
And the oplog-timestamp will not get updated anymore

**Comment by llvtt** [comment]
InvalidBSON is an Exception only raised by PyMongo... mongo-connector does
not raise this Exception and cannot do anything about it. No progress can
be made here until we understand what document causes this Exception to be
raised. If you can provide an example of a document that causes this
Exception, that would be extremely helpful.

**Comment by ilooch** [comment]
Hi, Luke
Thanks for your reply.
I find that the InvalidBSON is throw after another exception
> Call to <built-in function next> failed too many times in retry_until_ok
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/mongo_connector-2.2.dev0-py2.7.egg/mongo_connector/util.py", line 72, in retry_until_ok
    return func(*args, **kwargs)
  File "build/bdist.linux-x86_64/egg/pymongo/cursor.py", line 1090, in next
    if len(self.__data) or self._refresh():
  File "build/bdist.linux-x86_64/egg/pymongo/cursor.py", line 1022, in _refresh
    self.__uuid_subtype))
  File "build/bdist.linux-x86_64/egg/pymongo/cursor.py", line 958, in __send_message
    self.__compile_re)
  File "build/bdist.linux-x86_64/egg/pymongo/helpers.py", line 121, in _unpack_response
    compile_re)
  File "build/bdist.linux-x86_64/egg/bson/__init__.py", line 537, in decode_all
    tz_aware, uuid_subtype, compile_re))
  File "build/bdist.linux-x86_64/egg/bson/__init__.py", line 331, in _elements_to_dict
    data, position, as_class, tz_aware, uuid_subtype, compile_re)
  File "build/bdist.linux-x86_64/egg/bson/__init__.py", line 320, in _element_to_dict
    data, position, as_class, tz_aware, uuid_subtype, compile_re)
  File "build/bdist.linux-x86_64/egg/bson/__init__.py", line 159, in _get_object
    encoded, as_class, tz_aware, uuid_subtype, compile_re)
  File "build/bdist.linux-x86_64/egg/bson/__init__.py", line 331, in _elements_to_dict
    data, position, as_class, tz_aware, uuid_subtype, compile_re)
  File "build/bdist.linux-x86_64/egg/bson/__init__.py", line 320, in _element_to_dict
    data, position, as_class, tz_aware, uuid_subtype, compile_re)
  File "build/bdist.linux-x86_64/egg/bson/__init__.py", line 234, in _get_date
    dt = EPOCH_NAIVE + datetime.timedelta(seconds=seconds)
InvalidBSON: date value out of range

And the root exception raised by mongo-connector is
>2015-08-04 03:43:55,344 [ERROR] mongo_connector.oplog_manager:482 - OplogThread: caught exception during bulk upsert, re-upserting documents serially
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/mongo_connector-2.2.dev0-py2.7.egg/mongo_connector/oplog_manager.py", line 479, in upsert_all
    dm.bulk_upsert(docs_to_dump(namespace), mapped_ns, long_ts)
  File "/usr/local/lib/python2.7/dist-packages/mongo_connector-2.2.dev0-py2.7.egg/mongo_connector/util.py", line 32, in wrapped
    return f(*args, **kwargs)
  File "/usr/local/lib/python2.7/dist-packages/mongo_connector-2.2.dev0-py2.7.egg/mongo_connector/doc_managers/elastic_doc_manager.py", line 190, in bulk_upsert
    for ok, resp in responses:
  File "build/bdist.linux-x86_64/egg/elasticsearch/helpers/__init__.py", line 155, in streaming_bulk
    raise BulkIndexError('%i document(s) failed to index.' % len(errors), errors)
BulkIndexError: ('500 document(s) failed to index.', [{u'index': {u'status': 500, u'_type': u'mongodb_meta', u'_id': u'54939e414bf37bf33878817b', u'error': u'IndexCreationException[[mongodb_meta] failed to create index]; nested: NullPointerException; ', u'_index': u'mongodb_meta'}}, {u'index': {u'status': 500, u'_type': u'mongodb_meta', u'_id': u'549f44bd5f4adf2b692123d0', u'error': u'IndexCreationException[[mongodb_meta] failed to create index];  
...

There are a lot of this _u'IndexCreationException_ with different *_id*

**Comment by llvtt** [comment]
Normally, these Exceptions wouldn't be related. Does Python say that they
are (i.e., is there a phrase that says "caused by" in the Exception log)?
InvalidBSON is only raised for documents coming directly out of MongoDB-
this Exception indicates that there is some document with a Date that
doesn't make any sense to PyMongo. The BulkIndexError comes from
Elasticsearch, and it almost certainly indicates an entirely different
issue.

I can't tell what's causing the BulkIndexError on the Elasticsearch side
without seeing the Elasticsearch logs. Could you link to them here, please?

**Comment by joe-mojo** [comment]
I will look for a example of data that trigger the INvalidBSON exception — if I still have it.
In the mean time, I suspect this kinf of data to kill mongo-connector : date field thar are exported as {$date: { $numberLong: 12345... }}. I think values in the far future (after 2039) is suspicious too.

**Comment by behackett** [comment]
I think you are seeing this in PyMongo:

https://api.mongodb.org/python/current/faq.html#why-do-i-get-overflowerror-decoding-dates-stored-by-another-language-s-driver

Though I don't understand why the oplog would have timestamps with a year before 1 or after 9999. It's possible that there is a corrupt record in the oplog, or a bug in the specific version of MongoDB you are using. What version of MongoDB are you using?

**Comment by joe-mojo** [comment]
@behackett I'm using 2.4. (migration to 3 planned).
Years can be over 9999 in case of unfiltered type error (that was the case fo a long time in our application). But years above 9999 shouldn't be an issue. Any modern lib is not expected to overflow before the long timestamp epressing ms since 1970/01/01 00:00:00.000 UTC itself is overflown.
And more, I had some recorde before 9999 triggering the issue (2062 among other).

**Comment by ilooch** [comment]
@llvtt 
I reset the data and the issue was disappeared.
Will add the elasticsearch log once it appeared again.
Thanks.

**Comment by Sebmaster** [comment]
Got this error today and know what caused it:
MongoDB allows inserting arbitarly big dates, i.e. `ISODate("10217-06-06T00:00:00Z")` (not with the actual ISODate(str) constructor in the console though, that overflows). The python datetime lib only supports 4 digit years (until 9999), which is why it overflows when trying to convert eh long timestamp into a datetime instance.

**Comment by behackett** [comment]
> Years can be over 9999 in case of unfiltered type error (that was the case for a long time in our application). But years above 9999 shouldn't be an issue...

This is a limitation of the python datetime module, not PyMongo. The python datetime module requires that "year" be between 1 and 9999. The restriction is documented here:

https://docs.python.org/2/library/datetime.html#datetime.MINYEAR

There is nothing PyMongo (and therefore mongo-connector) can do to work around that limitation. On top of that, queries from MongoDB are returned and decoded in batches. If a document in the middle of the batch is corrupt, decoding of the batch fails. PyMongo can't currently decode the batch one at a time, which would likely be more expensive. So PyMongo can't "skip" the documents with invalid datetimes.

*However*, I noticed that the tracebacks listed in this ticket all indicate that PyMongo is being used in "pure python" mode. That is, it's C extensions are not installed. The C API of python's datetime module does not enforce the same restriction about max year. If you re-install PyMongo with its C extensions that *may* work around this issue. See the install docs for build dependencies here:

https://api.mongodb.org/python/current/installation.html#dependencies-for-installing-c-extensions-on-unix

**Comment by joe-mojo** [comment]
@behackett OK thx, and what about this fact : with a batch size of 1, mongo-connector still miss more than one document when a year > 9999 is found

**Comment by behackett** [comment]
You don't want a batch size of 1. That would mean every migrated change would be separate round trip to MongoDB, making mongo-connector unusably slow. Also, all current released versions of MongoDB have a weird limitation that requires batches to be of size 2 or greater (PyMongo automatically changes a batch size of 1 to 2 behind the scenes to work around that issue). That would explain your observation about batch size of 1.

The best long term solution is to fix the datetime fields in your application. You already know they are invalid. Might as well change the year to something PyMongo can handle.

**Comment by aherlihy** [comment]
Hello! I'm going to close this issue since we've found the source of the problem. However feel free to reopen or file a new issue if you are still experiencing trouble with the newer versions of mongo-connector (2.2) and/or the mongodb server (3.2). Thank you!

**Comment by vimuston** [comment]
We are experiencing this same issue. The problem is, that that the 'invalid' ISODate (year 9999+) is in a collection managed by another team/project, and mongo-connector is not even configured to follow that collection. This should not cause problems for the collection mongo-connector is following, since the two collections have nothing to do with each other.

**Comment by llvtt** [comment]
@vimuston The problem is that Python datetimes cannot represent dates in this range. Recently, a change in mongo-connector made it so that all documents from the oplog are inspected, but only the ones matching given namespace rules will actually be applied. The way to work around this is to make sure that the C extensions for PyMongo are installed; through the Python C API datetimes can be constructed to represent a larger range of dates. See [this comment](https://github.com/mongodb-labs/mongo-connector/issues/279#issuecomment-147862759). I'm also going to add this to the FAQ, since this comes up a lot.

**Comment by vimuston** [comment]
I understand the problem, I just don't understand why mongo-connector needs to look at any other entries on the oplog except the ones I have configured. There are other applications and teams using the same database, and I don't like it that the functionality of our application is now dependant on whether their data is correct and readable by Python.

**Comment by zvictor** [comment]
We are having `bson.errors.InvalidBSON: year is out of range` errors in our setup since we upgraded to `mongo-connector==2.4.1`. Downgrading it back to `mongo-connector==2.3` was the only way to solve the issue.

Log output:
```haskell
Traceback (most recent call last):
  File "/root/anaconda3/lib/python3.5/site-packages/mongo_connector/util.py", line 90, in wrapped
    func(*args, **kwargs)
  File "/root/anaconda3/lib/python3.5/site-packages/mongo_connector/oplog_manager.py", line 225, in run
    for n, entry in enumerate(cursor):
  File "/root/anaconda3/lib/python3.5/site-packages/pymongo/cursor.py", line 1090, in next
    if len(self.__data) or self._refresh():
  File "/root/anaconda3/lib/python3.5/site-packages/pymongo/cursor.py", line 1032, in _refresh
    self.__max_await_time_ms))
  File "/root/anaconda3/lib/python3.5/site-packages/pymongo/cursor.py", line 903, in __send_message
    codec_options=self.__codec_options)
  File "/root/anaconda3/lib/python3.5/site-packages/pymongo/helpers.py", line 142, in _unpack_response
    "data": bson.decode_all(response[20:], codec_options)}
bson.errors.InvalidBSON: year is out of range
2016-09-28 09:27:47,018 [ERROR] mongo_connector.connector:304 - MongoConnector: OplogThread <OplogThread(Thread-3, stopped 140371522139904)> unexpectedly stopped! Shutting down
2016-09-28 09:27:47,019 [INFO] mongo_connector.connector:362 - MongoConnector: Stopping all OplogThreads
2016-09-28 09:27:47,019 [DEBUG] mongo_connector.oplog_manager:377 - OplogThread: exiting due to join call.
```

Having the C extensions for PyMongo was not enough to solve it, as @llvtt just proposed.
```python
>>> import pymongo
>>> pymongo.has_c()
True
```

The process stops running after the error, even though `continueOnError: true` is set.

**Comment by ShaneHarvey** [comment]
Currently, `--continue-on-error` is only meant for errors that happen during the initial collection dump. It doesn't change the behavior while tailing the oplog.

Are you sure that the c extensions are installed for the same version of PyMongo and Python that mongo-connector is using?

**Comment by zvictor** [comment]
> Are you sure that the c extensions are installed for the same version of PyMongo and Python that mongo-connector is using?

positive.

**Comment by visonforcoding** [comment]
can`t catch this error ,then skip it ? i miss this error too. i feel very sad .

**Comment by uriinf** [comment]
Confirm that pymongo with C extension doesn't help with year > 9999.
My configuration:
```
Python 3.7.3 (default, Mar 27 2019, 23:44:26) 
[GCC 6.3.0 20170516] on linux
>>> import pymongo
>>> pymongo.has_c()
True
>>> pymongo.version
'3.8.0'
>>> from mongo_connector.constants import __version__
>>> __version__
'3.1.1'
```
And it still crashes with exception:
```ERROR] mongo_connector.util:109 - Fatal Exception
Traceback (most recent call last):
  File "/home/yurii/workspace/git/mongo-connector/mongo_connector/util.py", line 107, in wrapped
    func(*args, **kwargs)
  File "/home/yurii/workspace/git/mongo-connector/mongo_connector/oplog_manager.py", line 245, in run
    for n, entry in enumerate(cursor):
  File "/home/yurii/.local/lib/python3.7/site-packages/pymongo/cursor.py", line 1225, in next
    if len(self.__data) or self._refresh():
  File "/home/yurii/.local/lib/python3.7/site-packages/pymongo/cursor.py", line 1162, in _refresh
    self.__send_message(g)
  File "/home/yurii/.local/lib/python3.7/site-packages/pymongo/cursor.py", line 1006, in __send_message
    legacy_response=legacy_response, user_fields=user_fields)
  File "/home/yurii/.local/lib/python3.7/site-packages/pymongo/cursor.py", line 1097, in _unpack_response
    legacy_response)
  File "/home/yurii/.local/lib/python3.7/site-packages/pymongo/message.py", line 1432, in unpack_response
    return bson.decode_all(self.documents, codec_options)
bson.errors.InvalidBSON: year 10352 is out of range
```
I tried downgrade mongo-connector==2.3.0 accordingly to @zvictor [comment](https://github.com/yougov/mongo-connector/issues/279#issuecomment-250132247) but still with no success.

So only way worked for me is downgrading both - pymongo==2.9 and mongo-connector==2.3.0

Hope it will save some time for people who get this issue.
